File: backend/src/modules/users/user.model.ts
import { Schema, model, Document, Types } from 'mongoose';

export interface IUserPreferences {
  pinnedModules: string[];
  moduleOrder: string[];
  largeText: boolean;
}

export interface IUser extends Document {
  _id: Types.ObjectId;
  username: string;
  passwordHash: string;
  displayName: string;
  role: 'admin' | 'user';
  isActive: boolean;
  preferences: IUserPreferences;
  lastLoginAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

const userSchema = new Schema<IUser>(
  {
    username: {
      type: String,
      required: true,
      unique: true,
      lowercase: true,
      trim: true,
      index: true,
    },
    passwordHash: {
      type: String,
      required: true,
    },
    displayName: {
      type: String,
      required: true,
      default: 'Mẹ',
    },
    role: {
      type: String,
      enum: ['admin', 'user'],
      default: 'user',
    },
    isActive: {
      type: Boolean,
      default: true,
    },
    preferences: {
      pinnedModules: {
        type: [String],
        default: ['boat_receipts'],
      },
      moduleOrder: {
        type: [String],
        default: ['boat_receipts', 'games'],
      },
      largeText: {
        type: Boolean,
        default: true,
      },
    },
    lastLoginAt: {
      type: Date,
    },
  },
  {
    timestamps: true,
  }
);

export const UserModel = model<IUser>('User', userSchema, 'users');

File: frontend/lib/features/auth/models/user_model.dart
class UserPreferences {
  final List<String> pinnedModules;
  final List<String> moduleOrder;
  final bool largeText;

  UserPreferences({
    required this.pinnedModules,
    required this.moduleOrder,
    required this.largeText,
  });

  factory UserPreferences.fromJson(Map<String, dynamic> json) {
    return UserPreferences(
      pinnedModules: List<String>.from(json['pinnedModules'] ?? ['boat_receipts']),
      moduleOrder: List<String>.from(json['moduleOrder'] ?? ['boat_receipts', 'games']),
      largeText: json['largeText'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'pinnedModules': pinnedModules,
      'moduleOrder': moduleOrder,
      'largeText': largeText,
    };
  }
}

class UserModel {
  final String id;
  final String username;
  final String displayName;
  final String role;
  final UserPreferences preferences;

  UserModel({
    required this.id,
    required this.username,
    required this.displayName,
    required this.role,
    required this.preferences,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['id'] ?? json['_id'] ?? '',
      username: json['username'] ?? '',
      displayName: json['displayName'] ?? 'Mẹ',
      role: json['role'] ?? 'user',
      preferences: json['preferences'] != null
          ? UserPreferences.fromJson(json['preferences'])
          : UserPreferences(pinnedModules: ['boat_receipts'], moduleOrder: ['boat_receipts', 'games'], largeText: true),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'username': username,
      'displayName': displayName,
      'role': role,
      'preferences': preferences.toJson(),
    };
  }
}
